### PR TITLE
Added MMU tests

### PR DESCRIPTION
**Pull request by linkuri267**
Added three MMU tests. Two are for testing low address read/writes through MMU and one for testing mappings to high addresses. Changed Makefile, Makefile configuration, main.c and memory map accordingly. Tests run automatically at TRCH startup if Makefile configuration is set. Tests are defaulted to run currently. On test failure, panic() will be called.

### CODE REVIEW

File: plat/mem-map.h [hunk 1]
```diff
@@ -70,6 +70,10 @@
 #define RT_MMU_TEST_DATA_HI_1_ADDR      0x100010000
 #define RT_MMU_TEST_DATA_HI_SIZE            0x10000
 
+#define MMU_TEST_DATA_HI_0_ADDR        0x0500000000
+#define MMU_TEST_DATA_HI_1_ADDR        0x0500001000
+#define MMU_TEST_DATA_HI_SIZE		   	   0x001000
```
Comment (kdatta6) [inline plat/mem-map.h]: Looks like tabs are being used here, which misaligns the mem address from the position in the previous line.  As always, please use spaces to align.

File: plat/mem-map.h [hunk 2]
```diff
@@ -97,4 +101,8 @@
 #define RT_MMU_TEST_DATA_LO_ADDR         0xc3fff000
 #define RT_MMU_TEST_DATA_LO_SIZE           0x001000
 
+#define MMU_TEST_DATA_LO_0_ADDR 		 0xc2ffe000
+#define MMU_TEST_DATA_LO_1_ADDR		 	 0xc2fff000
+#define MMU_TEST_DATA_LO_SIZE		   	   0x001000
```
Comment (kdatta6) [inline plat/mem-map.h]: Please avoid tab usage when doing text alignment.

File: trch/main.c [hunk 3]
```diff
@@ -133,6 +133,24 @@ int main ( void )
     struct dma *trch_dma = NULL;
 #endif // !CONFIG_TRCH_DMA
 
+#if TEST_32_MMU_ACCESS_PHYSICAL
+    if(test_32_mmu_access_physical_mwr(MMU_TEST_DATA_LO_0_ADDR, MMU_TEST_DATA_LO_1_ADDR, MMU_TEST_DATA_LO_SIZE)){
+        panic("MMU 32-bit access test map-write-read");
+    }
+    if(test_32_mmu_access_physical_wmr(MMU_TEST_DATA_LO_0_ADDR, MMU_TEST_DATA_LO_1_ADDR, MMU_TEST_DATA_LO_SIZE)){
+        panic("MMU 32-bit access test write-map-read");
+    }
```
Comment (kdatta6) [inline trch/main.c]: Two minor formatting points in order to be consistent with the rest of the file.  First, keep a space between each "if" and the opening paren.  Second, remove the braces since the then clause is only a single line.

File: trch/main.c [hunk 4]
```diff
@@ -133,6 +133,24 @@ int main ( void )
...
+
+
```
Comment (kdatta6) [inline trch/main.c]: Remove this extra whitespace.

File: trch/main.c [hunk 5]
```diff
@@ -133,6 +133,24 @@ int main ( void )
...
+#endif
+
+#if TEST_MMU_MAPPING_SWAP
+    if(test_mmu_mapping_swap(MMU_TEST_DATA_LO_0_ADDR, MMU_TEST_DATA_LO_1_ADDR, MMU_TEST_DATA_HI_0_ADDR, MMU_TEST_DATA_LO_SIZE)){ //map argument 1 to argument 3 then argument 2 to argument 3
+        panic("MMU mapping swap");
+    }
+
```
Comment (kdatta6) [inline trch/main.c]: Same comments as above.

File: trch/tests/mmu.c [hunk 6]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+    struct mmu_context *trch_ctx;
+    struct mmu_stream *trch_stream;
+    struct balloc *ba;
+
+    uint32_t val = 0xbeeff00d;
+
+    mmu_32 = mmu_create("RTPS/TRCH->HPPS", RTPS_TRCH_TO_HPPS_SMMU_BASE);
+
+    if (!mmu_32)
+    return 1;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Please indent this line.

File: trch/tests/mmu.c [hunk 7]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+
+    if (*((uint32_t*)addr_to) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu 32-bit access physical map-write-read: read wrong value\n");
+        success = 0;
+    }
+
+    *((uint32_t*)addr_to) = 0;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: I think there are two problems with this line of code.  First, it will only work if the test is successful; if it's not, then you may be writing unsafely into memory.  One way to avoid this may be to enable the MMU again and then write to the address pointed to by addr_from.  Second, you want this test to ultimately leave the contents of memory unchanged- I think that's why you added this line.  To do that, you may want to check what the original contents of this memory address were before you changed it, save that info, and then restore it here.

File: trch/tests/mmu.c [hunk 8]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
     addr = (volatile uint32_t *)RT_MMU_TEST_DATA_HI_0_WIN_ADDR;
     val = 0xf00dbeef;
-    printf("%p <- %08x\r\n", addr, val);
+    printf("%p <- %08lx\r\n", addr, val);
     *addr = val;
 
     return 0;
 }
+
+int test_32_mmu_access_physical_mwr(uint32_t addr_from, uint32_t addr_to, unsigned mapping_sz){
```
Comment (kdatta6) [inline trch/tests/mmu.c]: The meaning of "mwr" is a bit unclear.  I think it means map-write-read, so it would be good just to mention that in the comments below.  In addition, it may be better to use the parameter names like "phy_write_addr" and "virt_read_addr" rather than the current ones.

File: trch/tests/mmu.c [hunk 9]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+cleanup_balloc:
+    mmu_destroy(mmu_32);
+
+    if (success) {
+        return 0;
+    }
+    return 1;
+}
+
+int test_32_mmu_access_physical_wmr(uint32_t addr_from, uint32_t addr_to, unsigned mapping_sz){
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Please mention in the comments below that "wmr" means map-write-read for clarity.  Also, same as above, it may be better to use more descriptive parameter names (e.g. "virt_read_addr", etc.)

File: trch/tests/mmu.c [hunk 10]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+    struct mmu_context *trch_ctx;
+    struct mmu_stream *trch_stream;
+    struct balloc *ba;
+
+    uint32_t val = 0xbeeff00d;
+
+    mmu_32 = mmu_create("RTPS/TRCH->HPPS", RTPS_TRCH_TO_HPPS_SMMU_BASE);
+
+    if (!mmu_32)
+    return 1;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Please indent this line.

File: trch/tests/mmu.c [hunk 11]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+
+    if (*((uint32_t*)addr_from) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu 32-bit access physical write-map-read: read wrong value\n");
+        success = 0;
+    }
+
+    *((uint32_t*)addr_from) = 0;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: See the comment written above for the analogous line in the previous function.

File: trch/tests/mmu.c [hunk 12]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+    struct mmu_context *trch_ctx;
+    struct mmu_stream *trch_stream;
+    struct balloc *ba;
+
+    uint32_t val = 0xbeeff00d;
+
+    mmu_32 = mmu_create("RTPS/TRCH->HPPS", RTPS_TRCH_TO_HPPS_SMMU_BASE);
+
+    if (!mmu_32)
+    return 1;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Indent this line.

File: trch/tests/mmu.c [hunk 13]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+
+    if (*((uint32_t*)addr_from_2) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu mapping swap: read wrong value %lx \n", *((uint32_t*)addr_from_2));
+        success = 0;
+    }
+
+    *((uint32_t*)addr_from_2) = 0;
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Same comment as for the analogous line in the previous two functions.

File: trch/tests/mmu.c [hunk 14]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+cleanup_balloc:
+    mmu_destroy(mmu_32);
+
+    if (success) {
+        return 0;
+    }
+    return 1;
+}
+
+int test_mmu_mapping_swap(uint32_t addr_from_1, uint32_t addr_from_2, uint64_t addr_to, unsigned mapping_sz){
```
Comment (kdatta6) [inline trch/tests/mmu.c]: Same comment as above- please use parameter names like "phy_read_addr", etc. rather than the current ones so that the names are more descriptive.

File: trch/tests/mmu.c [hunk 15]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+    //In this test,
+    //1. TRCH creates a mapping from addr_from_1 to addr_to
+    //2. TRCH enables MMU
+    //3. TRCH writes data to addr_from_1 (effectively writing to addr_to)
+    //4. TRCH unmaps the mapping created in step 1
+    //5. TRCH maps addr_from_2 to addr_r
+    //4. TRCH reads addr_from_2, checking if data is same as in step 3
```
Comment (kdatta6) [inline trch/tests/mmu.c]: This should be cleaned up a little.  Step 4 is repeated, and there is no "addr_r" in step 5.  Also, make sure the steps you've listed are accurate.

File: trch/tests/mmu.c [hunk 16]
```diff
@@ -18,13 +20,278 @@ int test_rt_mmu()
...
+    }
+
+    mmu_enable(mmu_32);
+
+    *((uint32_t*)addr_from_1) = val;
+
+    if (*((uint32_t*)addr_from_1) != val) {
+        printf("test mmu mapping swap: intial read failed\n");
+        goto cleanup_read;
+    }
```
Comment (kdatta6) [inline trch/tests/mmu.c]: I don't think this check is necessary.

File: trch-bl0/vectors.s.gen [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+exc12:
+    b exc12
+
+.thumb_func
+exc13:
+    b exc13
+
+.thumb_func
+exc14:
+    b exc14
```
Comment (kdatta6) [inline trch-bl0/vectors.s.gen]: This file (vectors.s.gen) is an automatically generated file and should not be included in the PR.  Please remove this file and confirm that it is still generated during the build.

File: trch/tests/mmu.c [hunk 18]
```diff
@@ -18,13 +20,282 @@ int test_rt_mmu()
...
     val = 0xf00dbeef;
-    printf("%p <- %08x\r\n", addr, val);
+    printf("%p <- %08lx\r\n", addr, val);
     *addr = val;
 
     return 0;
 }
+
+int test_32_mmu_access_physical_mwr(uint32_t virt_write_addr, uint32_t phy_read_addr, unsigned mapping_sz){
+    //map-read-write test
```
Comment (kdatta6) [inline trch/tests/mmu.c]: This should be "map-write-read test"

File: trch/tests/mmu.c [hunk 19]
```diff
@@ -18,13 +20,282 @@ int test_rt_mmu()
...
+
+    if (*((uint32_t*)virt_read_addr) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu 32-bit access physical write-map-read: read wrong value\n");
+        success = 0;
+    }
+
+    *((uint32_t*)virt_read_addr) = old_val;
```
Comment (kdatta6) [inline trch/tests/mmu.c:186]: I think the safest way to restore the memory contents would be to do the reverse of how you set old_val in the first place.  Specifically, it's probably best to first disable the MMU, then set the contents at address phy_write_addr to be equal to old_val.

File: trch/tests/mmu.c [hunk 20]
```diff
@@ -18,13 +20,282 @@ int test_rt_mmu()
...
+}
+
+int test_mmu_mapping_swap(uint32_t virt_write_addr, uint32_t virt_read_addr, uint64_t phy_addr, unsigned mapping_sz){
+    //In this test,
+    //1. TRCH creates a mapping from virt_write_addr to phy_addr
+    //2. TRCH enables MMU
+    //3. TRCH writes data to virt_write_addr (effectively writing to phy_addr)
+    //4. TRCH unmaps the mapping created in step 1
+    //5. TRCH maps virt_read_addr to phy_addr
+    //4. TRCH reads virt_read_addr, checking if data is same as in step 3
```
Comment (kdatta6) [inline trch/tests/mmu.c]: This should be line "6", not "4".

File: trch/tests/mmu.c [hunk 21]
```diff
@@ -18,13 +20,282 @@ int test_rt_mmu()
...
+
+    if (*((uint32_t*)virt_read_addr) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu mapping swap: read wrong value %lx \n", *((uint32_t*)virt_read_addr));
+        success = 0;
+    }
+
+    *((uint32_t*)virt_read_addr) = old_val;
```
Comment (kdatta6) [inline trch/tests/mmu.c:283]: I think the safest thing to do in order to restore the original memory contents is to set old_val using phy_addr before enabling the MMU (for the first time), then use it to restore the memory contents after disabling the MMU (for the last time), again using phy_addr.

Comment (linkuri267) [reply trch/tests/mmu.c:283]: Since the phy_adr is in DDR high, we cannot access it directly.
